### PR TITLE
snapcraft: introduce click-based YAML configuration file support

### PR DESCRIPTION
**Pull request by cjp256**
- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

**Comment by cjp256** [comment]
Requires #2781  and #2780

**Comment by cjp256** [comment]
Rebased.

This PR enables it for remote-build as a limited test case.

I tested it with:
echo -e "launchpad-user: cjp256\nlaunchpad-accept-public-upload: true" > ~/.config/snapcraft/config.yaml

Then running:
`snapcraft remote-build` without args works without prompts.

**Comment by cjp256** [comment]
On a related note, there is an option of utilizing Click's `auto_envvar_prefix` for command groups to enable coverage of environment variable mappings to parameters. https://click.palletsprojects.com/en/7.x/options/#values-from-environment-variables

It would incorporate the command name, e.g. `remote-build` for mapping environment variables, e.g. `SNAPCRAFT_REMOTE_BUILD_LAUNCHPAD_USER=cjp256`.

With the Click wrapper, `_click_load_config()`, we could rely on cmd_name to produce similar behavior if there is some foreseen conflicts among parameter names between commands.  Presumably we would short circuit the lifecycleci commands as the common case.

File: snapcraft/cli/remote.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
 from typing import List
 from xdg import BaseDirectory
 from . import echo
+from .config import enable_snapcraft_config_file
```
Comment (sergiusens) [inline snapcraft/cli/remote.py]: why isn't this from ._config ?

Comment (sergiusens) [reply snapcraft/cli/remote.py]: It is not meant to be used outside of the cli package, right?

Comment (cjp256) [reply snapcraft/cli/remote.py]: Correct, changed to _config.

**Comment by cjp256** [comment]
Enabled config file for all commands except legacy...

**Comment by cjp256** [comment]
Added info that config file is being used.  Added warning that config file is experimental.

```
$ snapcraft remote-build
Using config file: /home/ubuntu/.config/snapcraft/config.yaml
Configuration file is experimental and is subject to change.
snapcraft remote-build is experimental and is subject to change - use with caution.
```

**Comment by cjp256** [comment]
Rebased

**Comment by cjp256** [comment]
Accidentally forced push wrong branch over this. Fixed.

**Comment by cjp256** [comment]
Scaled it back to just lifecycle and remote commands.

File: snapcraft/cli/lifecycle.py [hunk 2]
```diff
@@ -262,6 +270,7 @@ def snap(directory, output, **kwargs):
 
 
 @lifecyclecli.command(cls=SnapcraftProjectCommand)
+@enable_snapcraft_config_file()
```
Comment (sergiusens) [inline snapcraft/cli/lifecycle.py]: pack shouldn't need this

Comment (cjp256) [reply snapcraft/cli/lifecycle.py]: Removed.

**Comment by cjp256** [comment]
rebased/squashed

**Pull request merged**